vignettes/na-notebook.Rmd

```diff
@@ -0,0 +1,510 @@
+    }
+  }
+  return(list(max_df = max_df, max_version = max_version))
+}
+```
+
+```{r}
+cov_nas <- readRDS("cov_nas.rds") # simulated data, not real
+paste0("geo_value: ", unique(cov_nas$geo_value))
+paste0("version: ", "2022-02-04")
+plot_signals(complete(cov_nas))
+```
+
+As a side note, notice how we use the `complete()` function here from the `tidyr` package. It is commonly used as a way to standardize our data as it ensures that all combinations of specified keys such as the time value and geos are present, even if this results in some missing values. If certain combinations are missing in the original data, `complete()` fills them in with rows containing NA values. 
+
+Here is small example:
+```{r}
+data <- tibble(
+  geo = c("CA", "CA", "NY"),
+  time = as.Date(c("2020-01-01", "2020-01-03", "2020-01-01")),
+  value = c(10, 20, 30)
+)
+
+data_completed <- data %>%
+  complete(
+    geo,
+    time = seq.Date(from = min(time), to = max(time), by = "day")
+  )
+
+print(data_completed)
+```
+
+
+Returning back, as we can see from both the plot and the output of our `plot_signals()` function that there are many gaps of NA values (13 in total), many of which with only a few days missing, but one gap with almost a year of missing values. For now we will focus on the gaps that are relatively small. These NAs have likely arisen due to reporting delays or inconsistencies in data collection from the hhs source.
+
+The process of filling in these NA values are known as imputation. Typically when imputing missing values, we use prior data in an attempt to predict what the signal would have been (there are exceptions as we will see later with the Google data).
+
+There are two different time scales we can impute on when analyzing `epi_archive`s. Imputation in time and imputation in version. Imputation in time fills in missing values of the signal based on previous values from previous times points of that signal. On the other hand, because our signals have multiple versions of the same signal, which arise from correcting previous recordings of the data, we can use these previous versions of the data and fill in the missing value for a day based on the value from previous versions. This is known as imputation in version. We holding constant the same day, but looking back at previous values that this day had in other versions of the data.
+
+To deal with these smaller NA gaps we will impute in time (as opposed to in version). Two of the most common and more simpler imputation schemes are the last observation carried forward (LOCF) method and a moving average imputation method. As suggested in the name of the method, to do LOCF imputation you simply use the previous available value for the missing value. Here is the corresponding code for LOCF.
+
+```{r locf_fn, warning=FALSE}
+impute_locf <- function(data) {
+  data %>%
+    group_by(geo_value) %>%
+    mutate(across(where(is.numeric), ~ zoo::na.locf(.x, na.rm = FALSE), .names = "{.col}_locf")) %>%
```

Comment:
todo: use `fill` from `tidyr` to do this instead, to focus on a smaller set of packages.  I guess we should probably include an `arrange` to make sure things are arranged by `time_value` within each `geo_value` as well.